standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HMSHandler.java

```diff
@@ -3633,21 +3633,46 @@ private Table getTableInternal(GetTableRequest getTableRequest) throws MetaExcep
   @Override
   public List<TableMeta> get_table_meta(String dbnames, String tblNames, List<String> tblTypes)
       throws MetaException, NoSuchObjectException {
+
     List<TableMeta> t = null;
+    List<TableMeta> finalT = new ArrayList<>();
     String[] parsedDbName = parseDbName(dbnames, conf);
     startTableFunction("get_table_metas", parsedDbName[CAT_NAME], parsedDbName[DB_NAME], tblNames);
     Exception ex = null;
     try {
       t = getMS().getTableMeta(parsedDbName[CAT_NAME], parsedDbName[DB_NAME], tblNames, tblTypes);
       t = FilterUtils.filterTableMetasIfEnabled(isServerFilterEnabled, filterHook,
           parsedDbName[CAT_NAME], parsedDbName[DB_NAME], t);
+
+      // metastore authorization : If any listeners are attached then for each tablemeta check if
+      // the database is readable, if so keep the tablemeta else remove it from the final list.
+      if(preListeners.size() > 0) {
+        Map<String, Boolean> databaseNames = new HashMap();
+        for (TableMeta tableMeta : t) {
+          String fullDbName = prependCatalogToDbName(parsedDbName[CAT_NAME], tableMeta.getDbName(), conf);
+          if (databaseNames.get(fullDbName) == null) {
+            boolean isExecptionThrown = false;
+            try {
+              fireReadDatabasePreEvent(fullDbName);
+            } catch (MetaException e) {
+              isExecptionThrown = true;
+            }
+            databaseNames.put(fullDbName, isExecptionThrown);
+          }
+          if (!databaseNames.get(fullDbName)) {
+            finalT.add(tableMeta);
+          }
+        }
+      } else {
+        finalT.addAll(t);
+      }
```

Comment:
I think instead of adding this block here ; you could move this content into a method; and follow the existing
```
t= someFilteringMethod(t, ...)
```

Reply:
ack

Reply:
ack